bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/discovery/GreeDeviceFinder.java

```diff
@@ -0,0 +1,193 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.gree.internal.discovery;
+
+import static org.openhab.binding.gree.internal.GreeBindingConstants.*;
+
+import java.io.IOException;
+import java.io.StringReader;
+import java.net.DatagramPacket;
+import java.net.DatagramSocket;
+import java.net.InetAddress;
+import java.net.SocketTimeoutException;
+import java.net.UnknownHostException;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.gree.internal.GreeCryptoUtil;
+import org.openhab.binding.gree.internal.GreeException;
+import org.openhab.binding.gree.internal.gson.GreeScanReponsePackDTO;
+import org.openhab.binding.gree.internal.gson.GreeScanRequestDTO;
+import org.openhab.binding.gree.internal.gson.GreeScanResponseDTO;
+import org.openhab.binding.gree.internal.handler.GreeAirDevice;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.stream.JsonReader;
+
+/**
+ * The GreeDeviceFinder provides functionality for searching for GREE Airconditioners on the network and keeping a list
+ * of found devices.
+ *
+ * @author John Cunha - Initial contribution
+ * @author Markus Michels - Refactoring, adapted to OH 2.5x
+ */
+@NonNullByDefault
+public class GreeDeviceFinder {
+    private final Logger logger = LoggerFactory.getLogger(GreeDeviceFinder.class);
+
+    protected final InetAddress ipAddress;
+    protected HashMap<String, GreeAirDevice> deviceTable = new HashMap<>();
+
+    public GreeDeviceFinder() {
+        ipAddress = InetAddress.getLoopbackAddress(); // dummy
+    }
+
+    public GreeDeviceFinder(String broadcastAddress) throws UnknownHostException {
+        ipAddress = InetAddress.getByName(broadcastAddress);
+    }
+
+    public void scan(Optional<DatagramSocket> socket, boolean scanNetwork) throws GreeException {
+        validateSocket(socket);
+        byte[] sendData = new byte[1024];
+        byte[] receiveData = new byte[1024];
+
+        // Send the Scan message
+        GreeScanRequestDTO scanGson = new GreeScanRequestDTO();
+        scanGson.t = "scan";
+
+        GsonBuilder gsonBuilder = new GsonBuilder();
+        Gson gson = gsonBuilder.create();
+        String scanReq = gson.toJson(scanGson);
+        sendData = scanReq.getBytes();
```

Comment:
You should specify the charset here

Reply:
you mean like this?
String scanReq = new String (gson.toJson(scanGson), "UTF-8")

Reply:
I mean like this:
```suggestion
        sendData = scanReq.getBytes(StandardCharsets.UTF_8);
```

Reply:
changed